src/pcre2_compile_cgroup.c

```diff
@@ -378,9 +333,197 @@ while (TRUE)
     break;
     }
 
+  PCRE2_ASSERT(size > 0);
+  return size;
+  }
+}
+
+
+/* Returns TRUE, if the capture is in the list. */
+
+static BOOL
+PRIV(compile_is_capture_checked)(uint32_t capture_id,
+  uint16_t *captures, uint16_t *captures_end)
+{
+while (captures < captures_end)
+  {
+  if (*captures == capture_id)
+    return TRUE;
+
+  captures++;
+  }
+
+return FALSE;
+}
+
+
+/*******************************************************
+*   Parse the arguments of scan substring operations   *
+********************************************************/
+
+/* This function parses the arguments of scan substring operations.
+
+Arguments:
+  pptr_start    points to the current parsed pattern pointer
+  offset        argument starting offset in the pattern
+  errorcodeptr  where to put an error code
+  cb            the compile block
+  lengthptr     NULL during the real compile phase
+                points to length accumulator during pre-compile phase
+
+Returns:        TRUE if OK, FALSE if not, error code set
+*/
+
+uint32_t *
+PRIV(compile_parse_scan_substr_args)(uint32_t *pptr,
+  int *errorcodeptr, compile_block *cb, PCRE2_SIZE *lengthptr)
+{
+uint16_t *captures;
+uint16_t *captures_end;
+PCRE2_SPTR name;
+named_group *ng;
+named_group *end = cb->named_groups + cb->names_found;
+BOOL all_found;
+size_t size;
+
+PCRE2_ASSERT(*pptr == META_OFFSET);
+size = PRIV(compile_process_capture_list)(pptr - 1, 0, errorcodeptr, cb);
+if (size == 0) return NULL;
+
+captures = (uint16_t*)cb->cx->memctl.malloc(size * sizeof(uint16_t), cb->cx->memctl.memory_data);
+if (captures == NULL)
+  {
+  *errorcodeptr = ERR21;
+  READPLUSOFFSET(cb->erroroffset, pptr);
+  return NULL;
+  }
+
+captures_end = captures;
+while (TRUE)
+  {
+  switch (META_CODE(*pptr))
+    {
+    case META_OFFSET:
+    pptr++;
+    SKIPOFFSET(pptr);
+    continue;
+
+    case META_CAPTURE_NAME:
+    ng = cb->named_groups + pptr[1];
+    PCRE2_ASSERT((ng->hash_dup & NAMED_GROUP_IS_DUPNAME) != 0);
+    pptr += 2;
+    name = ng->name;
+
+    all_found = TRUE;
+    do
+      {
+      if (ng->name != name) continue;
+
+      if (!PRIV(compile_is_capture_checked)(ng->number, captures, captures_end))
+        {
+        all_found = FALSE;
+        *captures_end++ = ng->number;
+        }
+      }
+    while (++ng < end);
+
+    if (!all_found)
+      {
+      *lengthptr += 1 + 2 * IMM2_SIZE;
+      continue;
+      }
+
+    pptr[-2] = META_CAPTURE_NUMBER;
+    pptr[-1] = 0;
+    continue;
+
+    case META_CAPTURE_NUMBER:
+    pptr += 2;
+    if (PRIV(compile_is_capture_checked)(pptr[-1], captures, captures_end))
+      {
+      pptr[-1] = 0;
+      continue;
+      }
```

Comment:
I see what you mean. It really isn't much code to check for duplicates.

At the moment, regex compiling is O(n^2) because we do several linear searches in various places. I'm planning to hunt down and kill them, to get rid of denial-of-service attacks.

(There was a post about this on the pcre2-dev mailing list a while back. Some Erlang users wanted to be able to bound the amount of time that PCRE2 blocks the main loop. Making the parser worst-case O(n logn) would be a nice promise to make to users about how bad the pathological cases can get.)

Would it be really pedantic and annoying to drop the duplicate search, or else make a temporary sorted copy and do it that way instead?

No users will ever enter duplicates anyway.

Reply:
I can use a bitset (each capture is one bit) as an alternative. The memory needed is (max_capture+7/8) bytes.